### PR TITLE
Adding TLS options validation as final sig-network review follow up

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind cleanup
/kind api-change

**What this PR does / why we need it**:
I found a comment from @khenidak that I'd missed from the original sig-network review suggesting more thorough validation here. This adds generic "AnnotationKey" and "AnnotationValue" types in hopes that they'll be useful for more than just TLS options in the future.

**Does this PR introduce a user-facing change?**:
```release-note
* TLS options keys are now subject to the same validation as Kubernetes annotations.
* TLS options values now have a max length of 1024 characters.
```

### CODE REVIEW

**Review by howardjohn — APPROVED** [review]
lgtm. Note your release note has "TCP" instead of "TLS"

File: apis/v1alpha2/shared_types.go
Comment (jpeach) [inline apis/v1alpha2/shared_types.go:384]: That probably prevents keys that are JSON or PEM objects, which seems quire useful for TLS config.

Comment (robscott) [reply apis/v1alpha2/shared_types.go:384]: My first thought is that those may be better served by a reference to a separate secret or config map depending on the use case. With that said, we could increase the limit here to something like 16k if that helps unblock some specific use cases.

For reference, [upstream validation](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/validation/objectmeta.go#L36) ensures that the entire size of the annotations struct is less than 256kB. It's difficult to translate that perfectly, but since we currently allow for a max of 100

File: apis/v1alpha2/shared_types.go
Comment (youngnick) [inline apis/v1alpha2/shared_types.go:378]:
```suggestion
// AnnotationValue is the value of an annotation in Gateway API. This is used for
```

**Comment by youngnick** [comment]
Great change, just needs the docstring for `AnnotationValue` updated, then LGTM.

File: apis/v1alpha2/shared_types.go
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: Ended up bumping this up to 4096, hoping that's sufficient for now. That matches our longest length validation and still leaves us room to expand if necessary.

File: apis/v1alpha2/shared_types.go
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: Updated, thanks!

**Comment by youngnick** [comment]
/lgtm

/hold
for @jpeach

**Pull request merged**
